apps/site/assets/ts/tnm/helpers/process-realtime-data.ts

```diff
@@ -0,0 +1,316 @@
+
+const findHeadsign = (
+  data: RouteWithStopsWithDirections[],
+  routeIndex: number,
+  stopIndex: number,
+  directionIndex: number,
+  headsign: string
+): number => {
+  const index = data[routeIndex].stops_with_directions[stopIndex].directions[
+    directionIndex
+  ].headsigns.findIndex(({ name }) => name === headsign);
+  return index === -1
+    ? data[routeIndex].stops_with_directions[stopIndex].directions[
+        directionIndex
+      ].headsigns.length
+    : index;
+};
+
+const setRoute = (
+  data: RouteWithStopsWithDirections[],
+  index: number,
+  route: EnhancedRoute
+): RouteWithStopsWithDirections[] =>
+  dotProp.set(
+    dotProp.set(data, `${index}.route`, route),
+    `${index}.stops_with_directions`,
+    []
+  );
+
+const setStop = (
+  data: RouteWithStopsWithDirections[],
+  routeIndex: number,
+  index: number,
+  stop: Stop,
+  distance: string
+): RouteWithStopsWithDirections[] => {
+  let nextData = dotProp.set(
+    data,
+    `${routeIndex}.stops_with_directions.${index}.stop`,
+    stop
+  );
+  nextData = dotProp.set(
+    nextData,
+    `${routeIndex}.stops_with_directions.${index}.distance`,
+    distance
+  );
+  nextData = dotProp.set(
+    nextData,
+    `${routeIndex}.stops_with_directions.${index}.directions`,
+    [{ headsigns: [], direction_id: 0 }, { headsigns: [], direction_id: 1 }]
+  );
+
+  return nextData;
+};
+
+const setHeadsigns = (
+  data: RouteWithStopsWithDirections[],
+  routeIndex: number,
+  stopIndex: number,
+  predictedScheduleByHeadsign: PredictedScheduleByHeadsign
+): RouteWithStopsWithDirections[] =>
+  Object.keys(predictedScheduleByHeadsign).reduce(
+    (accumulator: RouteWithStopsWithDirections[], headsign: string) => {
+      const { direction_id: directionId } = predictedScheduleByHeadsign[
+        headsign
+      ];
+      const nextData = accumulator;
+
+      const headsignIndex = findHeadsign(
+        nextData,
+        routeIndex,
+        stopIndex,
+        directionId,
+        headsign
+      );
+
+      return dotProp.set(
+        nextData,
+        `${routeIndex}.stops_with_directions.${stopIndex}.directions.${directionId}.headsigns.${headsignIndex}`,
+        buildHeadsign(headsign, predictedScheduleByHeadsign)
+      );
+    },
+    data
+  );
+
+const buildHeadsign = (
+  headsign: string,
+  predictedSchedulesByHeadsign: PredictedScheduleByHeadsign
+): Headsign => ({
+  name: headsign,
+  times: predictedSchedulesByHeadsign[headsign].predicted_schedules.map(
+    ({ prediction, schedule }) => {
+      const shortPrediction = prediction
+        ? ({
+            track: prediction.track,
+            time: prediction.time,
+            status
+          } as Prediction)
+        : null;
+      const scheduledTime = schedule ? schedule.time : null;
+      const delay = 0;
+
+      return {
+        prediction: shortPrediction,
+        scheduled_time: scheduledTime,
+        delay
+      } as PredictedOrScheduledTime;
+    }
+  ),
+
+  train_number: ""
```

Comment:
this train number would be schedule.trip.name, but we only would display it for CR
https://github.com/mbta/dotcom/blob/master/apps/site/assets/ts/schedule/components/schedule-finder/TableRow.tsx#L164

Reply:
https://github.com/mbta/dotcom/blob/master/apps/site/assets/ts/components/Headsign.tsx#L110 I think setting that value will just work